### DISCUSSION TITLE
Lookup Field Component Specs

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#233] [Maintainer discussions]
The lookup field specs add more detail to a component included in the design of [Add table from import](https://wiki.mathesar.org/design/process/documents/table-import-specs)

Please follow the [Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines)

Two alternatives are listed in the document, and I wanted to include both as I considered the two options during the design process. However, I feel that the second alternative has very few drawbacks. The feedback I'm looking for is potential scenarios that I might have missed or improvements that could increase usability or user experience in general.

Feedback Questions:
- Is this component appropriate for the described scenarios?
- Are there any risks of errors that might hinder the user experience? cause loss of data? create additional effort for the user to fix?
-----
Document
https://hackmd.io/@mathesar/rJzSXoxou
-----
@kgodey @pavish @mathemancer

**Comment by mathemancer** [comment]
No loss of data, but with option two, they might have a typo resulting in failure to find the schema, and inadvertently create the new schema if they're not paying attention.  They might end up with some tables in `my_schema` and other tables in `my_schama`, and it would be hard to rectify that, since we're not planning to have the ability to move tables between schemata in the MVP (or are we?)

I.e., an extra step for safety might be warranted in this case.

**Reply by kgodey** [reply]
I agree with @mathemancer here, creating a new schema involves a major change to the structure of our database, so I think it makes sense to confirm with the user that that's what they want to do.

**Reply by pavish** [reply]
I agree to this as well. However, I do not think that it is a component level decision.

For the component, I prefer option 2, as it is more ARIA friendly.

For the schema creation flow, I would prefer having an explicit message before creation, showcasing a list of all the changes that will happen on the DB, and asking for confirmation from the user. Thoughts?

Also, for the Import flow, I think having the table name and schema lookup field within the content pane, instead of a modal, since that will allow the user to edit it even after the file is uploaded. This was mentioned earlier on hackmd, just an additional reminder.

**Reply by kgodey** [reply]
I don't think we need to showcase a list of changes happening in the DB, that sounds like it would be intimidating to non-technical users. I think making it clear that they're creating a new schema should be enough. @pavish I might be misunderstanding what you mean, though, so maybe you could do a quick sketch of what you mean?

**Reply by pavish** [reply]
@kgodey I was thinking of something like this, after the upload completes, and before the table gets created:

![Group 1](https://user-images.githubusercontent.com/11850603/121930372-eb684b00-cd5f-11eb-95fb-b78b609237cd.png)

I understand that it might be intimidating for non-technical users, especially those not using their own databases. In that case, we could only show an additional explanatory text below the lookup field mentioning that the schema will get created.

**Comment by ghislaineguerin** [comment]
@kgodey @mathemancer I've been thinking about the requirement of creating schema from the lookup field and the potential consequences of errors, at it seems to me that the benefits of adding a new schema in this way, aren't necessarily strong. Should we allow users to only select existing schemas and have to create them instead in a different context that provides a better experience?
@kgodey Moving this outside of the modal might help with the above. 
I think @pavish is right about us needing to consider the UX of communicating DB changes when they are not so explicit.

**Reply by kgodey** [reply]
@ghislaineguerin I think moving schema creation to a different context makes sense. I created a design issue for this already: https://github.com/centerofci/mathesar/issues/170, we can figure the design out when we get to it. For the file import step, we can assume that we are picking only from existing schemas.

**Reply by kgodey** [reply]
I also agree with moving it outside of the modal, I thought we had already decided to do that.

**Reply by pavish** [reply]
I am still not convinced of moving the creation to a whole different context. We could move the schema selection and table name input from out of the modal to the import content pane, as discussed.

As a user, I would really appreciate being able to create the schema on the same page itself. I wouldn't want to move to a different page and get back here (Although I do understand that schema creation is a relatively rare operation). However, I do not have a strong opinion, I'm inclined to go with your recommendations.

**Reply by kgodey** [reply]
I do think we should try not to interrupt the user's flow when they're creating the database objects they want (people may want to create schemas while creating tables). However, part of our goal with Mathesar is to guide users into good ways of setting up databases, and we don't necessarily want people to think they need a new schema for every table. One way to think about schemas is as equivalent to Airtable Bases (not entirely accurate, but I think it works in this context) – all related tables should be in a single schema. So it seems like a good idea to me to get the users used to the idea of operating _within_ a schema, rather than picking a schema for every table.

@ghislaineguerin you probably have enough feedback to make a recommendation here.

**Comment by ghislaineguerin** [comment]
@kgodey @mathemancer @pavish 
The review summary was added to the specs here: https://wiki.mathesar.org/design/process/documents/lookup-field-component-specs.md
Please confirm after you read it and I will finalize this review.

**Reply by mathemancer** [reply]
I've read it; looks good.

**Reply by kgodey** [reply]
Looks good.

**Reply by pavish** [reply]
Looks good to me!
